GenericJSON/Subscripting.swift

```diff
@@ -0,0 +1,96 @@
+import Foundation
+
+public extension JSON {
+
+    /// Get or set item on given index in an array
+    ///
+    /// For out-of-bounds indexes, reading returns `nil` and writing does nothing.
+    /// For non-array values, reading returns `nil` and writing converts the value
+    /// into an empty array. When writing, `nil` values get converted to `JSON.null`.
+    subscript(index: Int) -> JSON? {
+
+        get {
+            // Reading from a non-array value returns `nil`
+            guard case .array(let arr) = self else { return nil }
+            // Reading from an out-of-bounds index returns `nil`
+            guard arr.indices.contains(index) else { return nil }
+            // Regular access
+            return arr[index]
+        }
+
+        set {
+            // Writing to a non-array value converts the value into an empty array.
+            // TODO: This is not really very helpful, can we do better?
+            guard case .array(var arr) = self else {
+                self = .array([])
```

Comment:
Yes, this would be a very strange behavior. It also doesn't match with the analogous subscript case for object values.
What i don't really like is that we are converting the value into an array, but we are leaving it empty, thus ignoring the new value completely.
What we could do is convert the value into an array that has all the values between 0 and index filled with null and the index-th set with new value.
Something like:
```suggestion
                self = .array(Array(repeating: JSON.null, count: index))
                self[index] = newValue
```
However, for this behavior to be consistent we would have to modify line 30 as well and have a write to an out-of-bounds index behave as well.

Reply:
I agree. Growing the array automatically would be closer to JavaScript array semantics, but JavaScript uses some kind of sparse arrays and the “filler” is `undefined`, not `null`:

```javascript
var a = []
a[10] = 1
a // [10: 1]
a[2] // undefined
```

So how about doing nothing both for string/int subscript writes if the value is not an object/array? As proposed above:

```swift
var json: JSON = true
json[0] = 1
json.bar = "foo"
// json is still just `true`
```

If people want to just overwrite the value no matter what, they can simply assign:

```swift
var json: JSON = true
json = [1]
json = ["bar": "foo"]
```

Or force-merge. This would work even if `json` is intially `JSON.true`, for example:

```swift
var json: JSON = ["foo": true]
json.merge(with: ["bar": "foo"])
```

We wouldn’t have any support for writing to arbitrary array indices, but I doubt people really need that in Swift. What do you think?